File: problems/p096.py
from time import perf_counter
import random
start = perf_counter()

grids = []
with open('p096_sudoku.txt', 'r') as f:
    grid = None
    for line in f:
        if line[0] == 'G':
            grid != None and grids.append(grid)
            grid = []
        else:
            grid.append([int(i) for i in line.strip()])
    grids.append(grid)


def possible(g, x, y):
    a, b = 3 * (x // 3), 3 * (y // 3)
    square = [g[a + i][b + j] for i in range(3) for j in range(3)]
    col = [g[i][y] for i in range(9)]
    row = g[x]
    return set(range(1, 10)) - set(row + col + square)

def sudoku_solve(G):
    n = len(G)
    blanks = [(i, j) for i in range(n) for j in range(n) if G[i][j] == 0]
    cycles = 0
    def dfs(idx, sol):
        nonlocal cycles
        cycles += 1

        if idx >= len(blanks):
            return sol

        a, b = blanks[idx]
        for i in possible(sol, a, b):
            sol[a][b] = i
            x = dfs(idx + 1, sol)
            if x != None:
                return x
            sol[a][b] = 0

        return None

    solution = dfs(0, G)
    if solution == None:
        print('Unable to find a solution')
    else:
        print('It took %d cycles' % cycles)

    return solution

ans = 0
for i, grid in enumerate(grids):
    print('Solving Grid #%d' % (i + 1), end='... ')
    g = sudoku_solve(grid)
    ans += g[0][0] * 100 + g[0][1] * 10 + g[0][2]

print('Solution:', ans)

end = perf_counter()
print(end - start, 'seconds to run')

File: problems/p096.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

int* possible(int **sol, int a, int b) {
    int seen[10] = {};
    int available = 9;
    int x, y;
    for (int i = 0; i < 9; i++) {
        x = sol[a][i];
        y = sol[i][b];
        if (x != 0 && seen[x] == 0)
            available -= 1;
        seen[x] = 1;
        if (y != 0 && seen[y] == 0)
            available -= 1;
        seen[y] = 1;
    }

    int q = 3 * (a / 3), w = 3 * (b / 3);
    for (int j = 0; j < 3; j++) {
        for (int k = 0; k < 3; k++) {
            int c = sol[q + j][w + k];
            if (c != 0 && seen[c] == 0)
                available -= 1;
            seen[c] = 1;
        }
    }

    int *candidates = malloc(sizeof(int) * (1 + available));
    candidates[0] = available;
    int idx = 1;
    for (int i = 1; i < 10; i++) {
        if (seen[i] == 0) {
            candidates[idx++] = i;
        }
    }

    return candidates;
}

int** dfs(int *cycles, int idx, int num_blanks, int **blanks, int **sol) {
    *cycles += 1;
    if (idx >= num_blanks) {
        return (int**) sol;
    }
    int a = blanks[idx][0], b = blanks[idx][1];
    int *candidates = possible(sol, a, b);
    for (int i = 1; i <= candidates[0]; i++) {
        sol[a][b] = candidates[i];
        int **x = dfs(cycles, idx + 1, num_blanks, blanks, sol);
        if (x != NULL) {
            free(candidates);
            return x;
        } else {
            sol[a][b] = 0;
        }
    }
    free(candidates);
    return NULL;
}

int** sudoku_solve(int G[9][9]) {
    int num_blanks = 0;
    for (int i = 0; i < 9; i++) {
        for (int j = 0; j < 9; j++) {
            if (G[i][j] == 0) {
                num_blanks += 1;
            }
        }
    }

    int **blanks = malloc(sizeof(int*) * num_blanks);
    int idx = 0;
    for (int i = 0; i < 9; i++) {
        for (int j = 0; j < 9; j++) {
            if (G[i][j] == 0) {
                blanks[idx] = malloc(sizeof(int) * 2);
                blanks[idx][0] = i;
                blanks[idx][1] = j;
                idx++;
            }
        }
    }

    int cycles = 0;

    int **sol = malloc(sizeof(int*) * 9);
    for (int i = 0; i < 9; i++) {
        sol[i] = malloc(sizeof(int) * 9);
        memcpy(sol[i], G[i], sizeof(int) * 9);
    }


    sol = dfs(&cycles, 0, num_blanks, blanks, sol);
    for (int i = 0; i < num_blanks; i++) {
        free(blanks[i]);
    }
    free(blanks);

    if (sol == NULL) {
        printf("Unable to find a solution.\n");
        return (int**) G;
    } else {
        printf("It took %d cycles\n", cycles);
        return sol;
    }

}

int main(int argc, char *argv[]) {
    clock_t start = clock();

    FILE *fp = fopen("p096_sudoku.txt", "r");
    if (fp == NULL) {
        printf("Error: unable to open file %s\n", "p096_sudoku.txt");
        return 1;
    }
    int grids[50][9][9];
    char buffer[11];
    int r = 0, g = 0;
    while (fgets(buffer, sizeof(buffer), fp) != NULL) {
        if (buffer[0] == 'G')
            continue;
        for (int i = 0; i < 9; i++) {
            grids[g][r][i] = buffer[i] - 48;
        }
        r = (r + 1) % 9;
        if (r == 0) {
            g += 1;
        }
    }
    fclose(fp);

    int ans = 0;
    for (int i = 0; i < 50; i++) {
        printf("Solving Grid #%d...", i + 1);
        int **g = sudoku_solve(grids[i]);
        ans += g[0][0] * 100 + g[0][1] * 10 + g[0][2];
        for (int i = 0; i < 9; i++) {
            free(g[i]);
        }
        free(g);
    }

    printf("Solution: %d\n", ans);


    clock_t end = clock();
    double time = (double) (end - start) / CLOCKS_PER_SEC;
    printf("%f seconds to run\n", time);
}
